From 3f7e0edbfcd70b6d4ed0fd97c8a9bdd549ce79fa Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:23:00 -0400
Subject: [x86] perf: Add Haswell specific transaction flag reporting

Message-id: <1408462094-14194-12-git-send-email-jolsa@redhat.com>
Patchwork-id: 88004
O-Subject: [PATCH RHEL7.1 BZ1131394 011/325] perf/x86: Add Haswell specific transaction flag reporting
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit a405bad5ad2086766ce320b16a56952e013327f8
Author: Andi Kleen <ak@linux.intel.com>
Date: Fri Sep 20 07:40:40 2013 -0700

description
===========
In the PEBS handler report the transaction flags using the new
generic transaction flags facility. Most of them come from
the "tsx_tuning" field in PEBSv2, but the abort code is derived
from the RAX register reported in the PEBS record.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kernel/cpu/perf_event_intel_ds.c b/arch/x86/kernel/cpu/perf_event_intel_ds.c
index d915996..e368e01 100644
--- a/arch/x86/kernel/cpu/perf_event_intel_ds.c
+++ b/arch/x86/kernel/cpu/perf_event_intel_ds.c
@@ -206,6 +206,8 @@ union hsw_tsx_tuning {
  u64     value;
 };
 
+#define PEBS_HSW_TSX_FLAGS 0xff00000000ULL
+
 void init_debug_store_on_cpu(int cpu)
 {
  struct debug_store *ds = per_cpu(cpu_hw_events, cpu).ds;
@@ -800,6 +802,16 @@ static inline u64 intel_hsw_weight(struct pebs_record_hsw *pebs)
  return 0;
 }
 
+static inline u64 intel_hsw_transaction(struct pebs_record_hsw *pebs)
+{
+ u64 txn = (pebs->tsx_tuning & PEBS_HSW_TSX_FLAGS) >> 32;
+
+ /* For RTM XABORTs also log the abort code from AX */
+ if ((txn & PERF_TXN_TRANSACTION) && (pebs->ax & 1))
+  txn |= ((pebs->ax >> 24) & 0xff) << PERF_TXN_ABORT_SHIFT;
+ return txn;
+}
+
 static void __intel_pmu_pebs_event(struct perf_event *event,
        struct pt_regs *iregs, void *__pebs)
 {
@@ -878,10 +890,14 @@ static void __intel_pmu_pebs_event(struct perf_event *event,
      x86_pmu.intel_cap.pebs_format >= 1)
   data.addr = pebs->dla;
 
- /* Only set the TSX weight when no memory weight was requested. */
- if ((event->attr.sample_type & PERF_SAMPLE_WEIGHT) && !fll &&
-     (x86_pmu.intel_cap.pebs_format >= 2))
-  data.weight = intel_hsw_weight(pebs);
+ if (x86_pmu.intel_cap.pebs_format >= 2) {
+  /* Only set the TSX weight when no memory weight. */
+  if ((event->attr.sample_type & PERF_SAMPLE_WEIGHT) && !fll)
+   data.weight = intel_hsw_weight(pebs);
+
+  if (event->attr.sample_type & PERF_SAMPLE_TRANSACTION)
+   data.txn = intel_hsw_transaction(pebs);
+ }
 
  if (has_branch_stack(event))
   data.br_stack = &cpuc->lbr_stack;
-- 
1.7.1